blocks/algorithms.py

```diff
@@ -0,0 +1,227 @@
+"""Training algorithms."""
+from abc import ABCMeta, abstractmethod
+from collections import OrderedDict
+
+import theano
+from theano import tensor
+
+from blocks.utils import update_instance
+from blocks.graph import ComputationGraph
+
+
+class TrainingAlgorithm(object):
+    """Base class for training algorithms.
+
+    A training algorithm object has a simple life-cycle.
+    First it is initialized by calling its :meth:`initialize` method.
+    At this stage, for instance, Theano functions can be compiled.
+    After that the :meth:`process_batch` method is repeatedly
+    called with a batch of training data as a parameter.
+
+    """
+    __metaclass__ = ABCMeta
+
+    @abstractmethod
+    def initialize(self):
+        """Initialize the training algorithm."""
+        pass
+
+    @abstractmethod
+    def process_batch(self, batch):
+        """Process a batch of training data.
+
+        Attributes
+        ----------
+        batch : dict
+            A dictionary of (source name, data) pairs.
+
+        """
+        pass
+
+
+class TheanoTrainingAlgorithm(TrainingAlgorithm):
+    """A training algorithm based on Theano functions.
+
+    Parameters
+    ----------
+    cost : Theano variable
+        The objective to be minimized.
+    params : list of Theano shared variables
+        The parameters to be tuned.
+
+    Attributes
+    ----------
+    updates : list of (shared variable, Theano expression) tuples
+        Updates to be done for every batch. It is required that the
+        updates are done using the old values of optimized parameters.
+    cost : Theano variable
+        The objective to be minimized.
+    params : list of Theano shared variables
+        The parameters to be tuned.
+
+    Notes
+    -----
+        Changing `updates` attribute or calling `add_updates` after
+        the `initialize` method is called will have no effect.
+
+    .. todo::
+
+       Ideally the parameters should be extracted from the `cost`
+       computation graph. But in order for this to work our graph
+       traversal routine should recurse into ScanOp's, which is currently
+       not the case.
```

Comment:
Even better: walk the graph and find all shared vars. A tag can be used to
indixate that a shared should not be trained. Lets not make the trainin
algo dependent on the brick concept. About the graph it has to know anyway.

15 sty 2015 03:13, "Bart van Merriënboer" <notifications@github.com>
napisał(a):
>
> In blocks/algorithms.py:
>
> > +    cost : Theano variable
> > +        The objective to be minimized.
> > +    params : list of Theano shared variables
> > +        The parameters to be tuned.
> > +
> > +    Notes
> > +    -----
> > +        Changing `updates` attribute or calling `add_updates` after
> > +        the `initialize` method is called will have no effect.
> > +
> > +    .. todo::
> > +
> > +       Ideally the parameters should be extracted from the `cost`
> > +       computation graph. But in order for this to work our graph
> > +       traversal routine should recurse into ScanOp's, which is
currently
> > +       not the case.
>
> Can't we walk the graph and collect all the bricks, and then query their
params attributes? This should also give us all the parameters, no?
>
> —
> Reply to this email directly or view it on GitHub.